src/main/java/io/vertx/mqtt/MqttEndpoint.java

```diff
@@ -0,0 +1,344 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.vertx.mqtt;
+
+import io.netty.handler.codec.mqtt.MqttConnectReturnCode;
+import io.netty.handler.codec.mqtt.MqttQoS;
+import io.vertx.codegen.annotations.CacheReturn;
+import io.vertx.codegen.annotations.Fluent;
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.mqtt.messages.MqttPublishMessage;
+import io.vertx.mqtt.messages.MqttSubscribeMessage;
+import io.vertx.mqtt.messages.MqttUnsubscribeMessage;
+
+import java.util.List;
+
+/**
+ * Represents an MQTT endpoint for point-to-point communication with the remote MQTT client
+ */
+@VertxGen
+public interface MqttEndpoint {
+
+  /**
+   * Close the endpoint, so the connection with remote MQTT client
+   */
+  void close();
+
+  /**
+   * Client identifier as provided by the remote MQTT client
+   *
+   * @return
+   */
+  @CacheReturn
+  String clientIdentifier();
+
+  /**
+   * Authentication information as provided by the remote MQTT client
+   *
+   * @return
+   */
+  @CacheReturn
+  MqttAuth auth();
+
+  /**
+   * Will information as provided by the remote MQTT client
+   *
+   * @return
+   */
+  @CacheReturn
+  MqttWill will();
+
+  /**
+   * Protocol version required by the remote MQTT client
+   *
+   * @return
+   */
+  @CacheReturn
+  int protocolVersion();
+
+  /**
+   * Protocol name provided by the remote MQTT client
+   *
+   * @return
+   */
+  @CacheReturn
+  String protocolName();
+
+  /**
+   * If clean session is requested by the remote MQTT client
+   *
+   * @return
+   */
+  @CacheReturn
+  boolean isCleanSession();
+
+  /**
+   * Keep alive timeout (in seconds) specified by the remote MQTT client
+   *
+   * @return
+   */
+  @CacheReturn
+  int keepAliveTimeSeconds();
+
+  /**
+   * Message identifier used for last published message
+   *
+   * @return
+   */
+  @CacheReturn
+  int lastMessageId();
+
+  /**
+   * Enable/disable subscription/unsubscription requests auto acknowledge
+   *
+   * @param isSubscriptionAutoAck auto acknowledge status
+   */
+  void subscriptionAutoAck(boolean isSubscriptionAutoAck);
+
+  /**
+   * Return auto acknowledge status for subscription/unsubscription requests
+   *
+   * @return
+   */
+  boolean isSubscriptionAutoAck();
+
+  /**
+   * Enable/disable publishing (in/out) auto acknowledge
+   *
+   * @param isPublishAutoAck auto acknowledge status
+   */
+  void publishAutoAck(boolean isPublishAutoAck);
+
+  /**
+   * Return auto acknowledge status for publishing (in/out)
+   *
+   * @return
```

Comment:
@vietj sorry ... what does it mean ? what should be the format for a comment like this ?